clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -117,6 +118,17 @@ public int validBytes() {
         return downConvert(batches(), toMagic, firstOffset, time);
     }
 
+    /**
+     * Validates the header of the next batch and returns batch size. `buffer` should
+     * contain at least the header up to the size field. If buffer contains the magic
+     * byte, it is validated.
+     * @return next batch size including LOG_OVERHEAD
+     * @throws CorruptRecordException if record size or magic is invalid
+     */
+    public int nextBatchSize() {
+        return new ByteBufferLogInputStream(buffer, Integer.MAX_VALUE).nextBatchSize();
```

Comment:
Should we be duplicating the buffer?

Reply:
We dont change the position in the buffer in `nextBatchSize`, so `duplicate()` is not required? The unit test verifies that position is not changed.

`nextBatch()` returns null if the buffer is not large enough to hold the batch. `nextBatchSize` returns the batch size as long as the header is present, so we can allocate the buffer based on this size. Have I misunderstood the comment?